src/elona/foobar_map/map_loader.cpp

```diff
@@ -0,0 +1,210 @@
+#include "map_loader.hpp"
+#include "../../util/range.hpp"
+
+#include <iostream>
+
+namespace elona
+{
+namespace fmp
+{
+
+FoobarMap MapLoader::load(const fs::path& map_file)
+{
+    if (!fs::exists(map_file))
+    {
+        throw std::runtime_error(
+            "Map file does not exist: " +
+            filepathutil::make_preferred_path_in_utf8(map_file));
+    }
+
+    _map = {};
+
+    _in = std::make_unique<zstr::ifstream>(filepathutil::make_preferred_path_in_utf8(map_file));
+
+    char header[4];
+    _in->read(header, 4);
+
+    if (strncmp(header, "FMP ", 4) != 0)
+    {
+        throw std::runtime_error("Bad header");
+    }
+
+    _map.version = read_int();
+
+    int mod_count = read_int();
+    for (auto _i : range::iota<int>(mod_count))
+    {
+        _map.mods.emplace_back(read_string());
+    }
+
+    int mapping_count = read_int();
+    for (auto _i : range::iota<int>(mapping_count))
+    {
+        std::string name = read_string();
+        int id = read_int();
+        _map.ids_to_names[id] = name;
+    }
+
+    _map.props = read_properties();
+
+    _map.width = read_int();
+    _map.height = read_int();
+
+    _map.tiles = read_tiles();
+
+    _map.layers = read_layers();
+
+    return _map;
+}
+
+std::string MapLoader::read_string()
```

Comment:
Currently we don't want to embed NUL character in a string at all, but it cannot be said that we never want to.